### PR TITLE
Add a Sparse Merkle Tree library

### PR DESCRIPTION
**Pull request by KyrylR**
- [ ] Since this PR suggests a **bug fix**, the tests have been added and the coverage is 100%.
- [x] Since this PR introduces a **new feature**, the update has been discussed in an Issue or with the team.
- [ ] This PR is just a **minor change**, like a typo fix.

---

### CODE REVIEW

File: contracts/libs/data-structures/SparseMerkleTree.sol [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+        MIDDLE
+    }
+
+    struct SMT {
+        mapping(uint256 => Node) nodes;
+        uint64 merkleRootId;
+        uint64 maxDepth;
+        uint64 nodesCount;
+        bool isInitialized;
+        bool isCustomHasherSet;
```
Comment (Arvolear) [inline contracts/libs/data-structures/SparseMerkleTree.sol:224]: Can we assume that the tree is initialized if maxDepth > 0?

File: contracts/libs/data-structures/SparseMerkleTree.sol [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+        assembly {
+            mstore(0, a)
+            mstore(0x20, b)
+
+            result := keccak256(0, 0x40)
+        }
+    }
+
+    function _hash3(bytes32 a, bytes32 b, bytes32 c) private pure returns (bytes32 result) {
+        assembly {
```
Comment (Arvolear) [inline contracts/libs/data-structures/SparseMerkleTree.sol:513]: Maybe add a dev comment stating "we are not updating the free memory pointer here due to the temporary nature of hash arguments"

File: contracts/libs/data-structures/SparseMerkleTree.sol [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+
+            mstore(free_ptr, a)
+            mstore(add(free_ptr, 0x20), b)
+            mstore(add(free_ptr, 0x40), c)
+
+            result := keccak256(free_ptr, 0x60)
+        }
+    }
+
+    function _root(SMT storage tree) private view onlyInitialized(tree) returns (bytes32) {
```
Comment (Arvolear) [inline contracts/libs/data-structures/SparseMerkleTree.sol:524]: Let's not revert in view functions

File: contracts/libs/data-structures/SparseMerkleTree.sol [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+        function(bytes32, bytes32, bytes32) view returns (bytes32) hash3;
+    }
+
+    struct Node {
+        NodeType nodeType;
+        uint64 childLeft;
+        uint64 childRight;
+        bytes32 nodeHash;
+        bytes32 index;
+        bytes32 value;
```
Comment (Arvolear) [inline contracts/libs/data-structures/SparseMerkleTree.sol:235]: Can we change the naming a bit? To `key` and `value `

File: contracts/libs/data-structures/SparseMerkleTree.sol [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+                nodeHash: HASH_ZERO,
+                index: HASH_ZERO,
+                value: HASH_ZERO
+            });
+        }
+
+        return _setNode(tree, newNodeMiddle_);
+    }
+
+    function _setNode(SMT storage tree, Node memory node_) private returns (uint256) {
```
Comment (Arvolear) [inline contracts/libs/data-structures/SparseMerkleTree.sol:478]: Add a comment that weak hash functions will override the old nodes?

File: test/libs/data-structures/SparseMerkleTree.test.ts [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+      const onchainProof = await merkleTree.getUintProof(5n);
+
+      expect(onchainProof.auxIndex).to.equal(7n);
+      expect(onchainProof.auxValue).to.equal(1n);
+      expect(onchainProof.auxExistence).to.equal(true);
+      expect(onchainProof.existence).to.equal(false);
+    });
+
+    it("should reset the value of the node", async () => {
```
Comment (Arvolear) [inline test/libs/data-structures/SparseMerkleTree.test.ts:183]: Does original SMT do that?

Comment (KyrylR) [reply test/libs/data-structures/SparseMerkleTree.test.ts:183]: Yep, it does with a lot of asserts:

1. Initial call: https://github.com/iden3/contracts/blob/8815bf53e989311d94301f4c513436c1ff776911/contracts/lib/SmtLib.sol#L463
2. Asserts: https://github.com/iden3/contracts/blob/8815bf53e989311d94301f4c513436c1ff776911/contracts/lib/SmtLib.sol#L553

File: contracts/libs/data-structures/SparseMerkleTree.sol [hunk 7]
```diff
@@ -0,0 +1,935 @@
...
+        SMT _tree;
+    }
+
+    /**
+     * @notice The function to initialize the Merkle tree.
+     * Under the hood it sets the maximum depth of the Merkle tree, therefore can be considered
+     * alias function for the `setMaxDepth`.
+     *
+     * @param tree self.
+     * @param maxDepth_ The max depth of the Merkle tree. Depth must be 0.
```
Comment (RuslanProgrammer) [inline contracts/libs/data-structures/SparseMerkleTree.sol:70]: I think the depth of the tree should NOT be 0

File: contracts/libs/data-structures/SparseMerkleTree.sol [hunk 8]
```diff
@@ -1,6 +1,55 @@
...
+ *
+ * It aims to provide more gas-efficient operations for the Sparse Merkle Tree while offering flexibility
+ * in using different types of keys and values.
+ *
+ * The main differences from the original implementation include:
+ * - Optimized storage usage to reduce the number of storage slots.
+ * - Added the ability to set custom hash functions.
+ * - Removed methods and associated storage for managing the tree root's history.
+ *
+ * Gas usage for adding (addBytes32) 16,001 leaves is detailed below:
```
Comment (Arvolear) [inline contracts/libs/data-structures/SparseMerkleTree.sol:19]: For the tree of size 80?

**Review by Arvolear — APPROVED** [review]
LGTM!

**Pull request merged**